# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [2]:
train_df = pd.read_csv("clustered_co2.csv")
train_df.sample(5)

Unnamed: 0           Make                       Model    Vehicle Class  \
3262        3484       CADILLAC                     XTS AWD        FULL-SIZE   
4048        4322         TOYOTA             YARIS HATCHBACK          COMPACT   
1297        1328      CHEVROLET  EXPRESS 3500 PASSENGER FFV  VAN - PASSENGER   
5434        6172  MERCEDES-BENZ  AMG GLC 63 S 4MATIC+ Coupe      SUV - SMALL   
1536        1632       INFINITI                     Q70 AWD         MID-SIZE   

      Engine Size(L)  Cylinders Transmission Fuel Type  \
3262             3.6          6          AS6         X   
4048             1.5          4           M5         X   
1297             6.0          8           A6         X   
5434             4.0          8           A9         Z   
1536             5.6          8          AS7         Z   

      Fuel Consumption City (L/100 km)  Fuel Consumption Hwy (L/100 km)  \
3262                              13.7                              9.0   
4048                               7.8                              6.5   
1297                              22.2                             15.1   
5434                              15.0                             10.9   
1536                              15.0                             10.2   

      Fuel Consumption Comb (L/100 km)  Fuel Consumption Comb (mpg)  \
3262                              11.6                           24   
4048                               7.2                           39   
1297                              19.0                           15   
5434                              13.2                           21   
1536                              12.9                           22   

      CO2 Emissions(g/km)  Outlier  cluster  
3262                  272        1        0  
4048                  168        1        0  
1297                  437       -1        0  
5434                  309        1        1  
1536                  297        1        1

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [3]:
categorical_features = train_df.select_dtypes(include=['object']).columns
X = train_df.drop(columns=['cluster', *categorical_features])
y = train_df['cluster']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=12)

In [4]:
# menghitung panjang/jumlah data
print("Jumlah data: ",len(X))
# menghitung panjang/jumlah data pada x_test
print("Jumlah data latih: ",len(X_train))
# menghitung panjang/jumlah data pada x_test
print("Jumlah data test: ",len(X_test))

Jumlah data:  6282
Jumlah data latih:  5025
Jumlah data test:  1257


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

### random forest

In [5]:
RF_model = RandomForestClassifier()
RF_model.fit(X_train, y_train)

RF_y_train_pred = RF_model.predict(X_train)
RF_y_test_pred = RF_model.predict(X_test)

First, submitter will try to use random forest algorithm

### logistic regression

In [6]:
LR_model = LogisticRegression(max_iter=1000)
LR_model.fit(X_train, y_train)

LR_y_train_pred = LR_model.predict(X_train)
LR_y_test_pred = LR_model.predict(X_test)

/home/mendo/projects/python/ml_co2/venv/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


submitter also try logistic regression

### Decission Tree

In [7]:
DT_model = DecisionTreeClassifier()
DT_model.fit(X_train, y_train)

DT_y_train_pred = DT_model.predict(X_train)
DT_y_test_pred = DT_model.predict(X_test)

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

### random forest

In [8]:

print("Training Set Classification Report:")
print(classification_report(y_train, RF_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, RF_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      2544
           1       1.00      1.00      1.00      2213
           2       1.00      1.00      1.00       268

    accuracy                           1.00      5025
   macro avg       1.00      1.00      1.00      5025
weighted avg       1.00      1.00      1.00      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.94      0.93       643
           1       0.92      0.91      0.91       552
           2       1.00      1.00      1.00        62

    accuracy                           0.93      1257
   macro avg       0.95      0.95      0.95      1257
weighted avg       0.93      0.93      0.93      1257



evaluation results for the Random Forest algorithm shows that this model have the average as follow: accuracy of 90%, recall of 95% and f1-score of 95% as well.

### logistic regression

In [9]:
print("Training Set Classification Report:")
print(classification_report(y_train, LR_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, LR_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.73      0.69      2544
           1       0.64      0.56      0.60      2213
           2       1.00      1.00      1.00       268

    accuracy                           0.67      5025
   macro avg       0.76      0.76      0.76      5025
weighted avg       0.67      0.67      0.66      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.65      0.71      0.68       643
           1       0.62      0.55      0.58       552
           2       1.00      1.00      1.00        62

    accuracy                           0.65      1257
   macro avg       0.76      0.75      0.75      1257
weighted avg       0.65      0.65      0.65      1257



### Decission Tree

In [10]:
print("Training Set Classification Report:")
print(classification_report(y_train, DT_y_train_pred))
print( '-' * 50)
print("Test Set Classification Report:")
print(classification_report(y_test, DT_y_test_pred))

Training Set Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      2544
           1       1.00      1.00      1.00      2213
           2       1.00      1.00      1.00       268

    accuracy                           1.00      5025
   macro avg       1.00      1.00      1.00      5025
weighted avg       1.00      1.00      1.00      5025

--------------------------------------------------
Test Set Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.89      0.90       643
           1       0.87      0.90      0.89       552
           2       1.00      1.00      1.00        62

    accuracy                           0.90      1257
   macro avg       0.93      0.93      0.93      1257
weighted avg       0.90      0.90      0.90      1257



## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [11]:
#Type your code here

## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [12]:
#Type your code here

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

submitter applying three different algorithm Random Forest, Logistic Regression and Decission Tree. Found that Random Forest and Decission Tree have perfect evaluation score on training set while logistic regression have the worst of three.
result shows that random forest have a better evaluation score overall by 95% average for each recall, precission, and f1-score followed by decission tree at 93% each. Thus we can use either Decission Tree or Random Forest